File: src/ecs/energy.ts
import { clamp } from '@/lib/math';
import type { DroneEntity } from '@/ecs/world';

const EPSILON = 1e-4;

/**
 * Computes the energy availability fraction for a drone based on its battery level and throttle floor.
 * If the battery is depleted, the fraction will be limited by the throttle floor.
 *
 * @param drone - The drone entity.
 * @param throttleFloor - The minimum performance factor (0-1) when battery is low.
 * @returns A value between throttleFloor and 1 representing the energy availability factor.
 */
export const computeDroneEnergyFraction = (drone: DroneEntity, throttleFloor: number) => {
  if (drone.maxBattery <= 0) {
    return 1;
  }
  const normalized = drone.battery / drone.maxBattery;
  if (!Number.isFinite(normalized)) {
    return 1;
  }
  return clamp(normalized, throttleFloor, 1);
};

/**
 * Consumes energy from a drone's battery based on the requested rate and delta time.
 * Calculates the actual consumption and the resulting performance fraction.
 *
 * @param drone - The drone entity to consume energy from.
 * @param dt - Delta time in seconds.
 * @param throttleFloor - The minimum performance factor.
 * @param rate - The energy consumption rate per second.
 * @returns An object containing the effective performance `fraction` and the `consumed` energy amount.
 */
export const consumeDroneEnergy = (
  drone: DroneEntity,
  dt: number,
  throttleFloor: number,
  rate: number,
) => {
  const fraction = computeDroneEnergyFraction(drone, throttleFloor);
  if (dt <= 0 || rate <= 0) {
    drone.charging = false;
    return { fraction, consumed: 0 };
  }
  const consumed = Math.min(drone.battery, Math.max(0, rate * dt * fraction));
  if (consumed > 0) {
    drone.battery = Math.max(0, drone.battery - consumed);
    if (drone.battery <= EPSILON) {
      drone.battery = 0;
    }
  }
  drone.charging = false;
  return { fraction, consumed };
};

File: rust-engine/src/systems/energy.rs
use crate::constants::DRONE_MAX_BATTERY;

pub fn compute_drone_energy_fraction(battery: f32, throttle_floor: f32) -> f32 {
    if DRONE_MAX_BATTERY <= 0.0 {
        return 1.0;
    }
    let normalized = battery / DRONE_MAX_BATTERY;
    if !normalized.is_finite() {
        return 1.0;
    }
    normalized.clamp(throttle_floor, 1.0)
}

pub struct EnergyConsumption {
    pub fraction: f32,
    pub consumed: f32,
}

pub fn consume_drone_energy(
    battery: &mut f32,
    dt: f32,
    throttle_floor: f32,
    rate: f32,
) -> EnergyConsumption {
    let fraction = compute_drone_energy_fraction(*battery, throttle_floor);
    if dt <= 0.0 || rate <= 0.0 {
        return EnergyConsumption {
            fraction,
            consumed: 0.0,
        };
    }
    let consumed = (*battery).min((rate * dt * fraction).max(0.0));
    if consumed > 0.0 {
        *battery = (*battery - consumed).max(0.0);
        if *battery <= 1e-4 {
            *battery = 0.0;
        }
    }
    EnergyConsumption { fraction, consumed }
}
